### PR TITLE
[Pipeline](runtime filter) Support runtime filters on pipeline engine

### PR DESCRIPTION
**Pull request by Gabriel39**
# Proposed changes

Issue Number: close #xxx

## Problem summary

Describe your changes.

## Checklist(Required)

1. Does it affect the original behavior: 
    - [ ] Yes
    - [ ] No
    - [ ] I don't know
2. Has unit tests been added:
    - [ ] Yes
    - [ ] No
    - [ ] No Need
3. Has document been added or modified:
    - [ ] Yes
    - [ ] No
    - [ ] No Need
4. Does it need to update dependencies:
    - [ ] Yes
    - [ ] No
5. Are there any changes that cannot be rolled back:
    - [ ] Yes (If Yes, please explain WHY)
    - [ ] No

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 34.98 seconds
 load time: 439 seconds
 storage size: 17123356790 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20221213135223_clickbench_pr_62752.html

File: be/src/exec/exec_node.h [hunk 1]
```diff
@@ -96,6 +96,8 @@ class ExecNode {
     // so only vectorized exec node need to impl
     virtual Status alloc_resource(RuntimeState* state);
 
+    virtual bool runtime_filters_are_ready_or_timeout() { return true; }
```
Comment (HappenLee) [inline be/src/exec/exec_node.h:99]: better virtual function in `new_scan_node`

File: be/src/exprs/runtime_filter.cpp [hunk 2]
```diff
@@ -1239,29 +1240,95 @@ Status IRuntimeFilter::get_prepared_vexprs(std::vector<doris::vectorized::VExpr*
...
     int64_t wait_times_ms = _wrapper->get_real_type() == RuntimeFilterType::BITMAP_FILTER
                                     ? _state->query_options().query_timeout
                                     : _state->runtime_filter_wait_time_ms();
-    std::unique_lock<std::mutex> lock(_inner_mutex);
-    if (!_is_ready) {
-        int64_t ms_since_registration = MonotonicMillis() - registration_time_;
-        int64_t ms_remaining = wait_times_ms - ms_since_registration;
-        if (ms_remaining <= 0) {
-            return _is_ready;
+    if (_state->enable_pipeline_exec() &&
```
Comment (HappenLee) [inline be/src/exprs/runtime_filter.cpp:1247]: better code struct：
```
if (_state->enable_pipeline_exec()) {
   if (_rf_state_atomic.load() == xxx) { 
   } 
   if ()
} else {
}
```

File: be/src/exprs/runtime_filter.cpp [hunk 3]
```diff
@@ -1239,29 +1240,95 @@ Status IRuntimeFilter::get_prepared_vexprs(std::vector<doris::vectorized::VExpr*
...
+    if (!_state->enable_pipeline_exec()) {
+        _rf_state = RuntimeFilterState::TIME_OUT;
+        return true;
+    } else if (is_ready()) {
+        if (cur_state == RuntimeFilterState::NOT_READY) {
+            _profile->add_info_string("EffectTime", std::to_string(ms_since_registration) + " ms");
+        }
+        return true;
+    } else {
+        if (cur_state == RuntimeFilterState::NOT_READY) {
```
Comment (HappenLee) [inline be/src/exprs/runtime_filter.cpp:1297]: the logic here is wired? rethink?

File: be/src/pipeline/exec/scan_operator.h [hunk 4]
```diff
@@ -42,9 +42,7 @@ class ScanOperator : public SourceOperator<ScanOperatorBuilder> {
 
     bool is_pending_finish() const override;
 
-    Status open(RuntimeState* state) override;
-
-    Status close(RuntimeState* state) override;
```
Comment (HappenLee) [inline be/src/pipeline/exec/scan_operator.h:47]: why delete the code, other scan_node, like odbc, jdbc may need the function.

File: be/src/vec/exec/scan/vscan_node.cpp [hunk 5]
```diff
@@ -84,13 +84,22 @@ Status VScanNode::prepare(RuntimeState* state) {
...
+    if (_opened) {
+        return Status::OK();
+    }
+    SCOPED_CONSUME_MEM_TRACKER(mem_tracker_growh());
+    _input_tuple_desc = state->desc_tbl().get_tuple_descriptor(_input_tuple_id);
+    _output_tuple_desc = state->desc_tbl().get_tuple_descriptor(_output_tuple_id);
+    START_AND_SCOPE_SPAN(state->get_tracer(), span, "VScanNode::alloc_resource");
+    SCOPED_TIMER(_runtime_profile->total_time_counter());
+    RETURN_IF_ERROR(ExecNode::alloc_resource(state));
     RETURN_IF_ERROR(_acquire_runtime_filter());
```
Comment (HappenLee) [inline be/src/vec/exec/scan/vscan_node.cpp:103]: the function seems need to pass value `wait`?

File: be/src/pipeline/pipeline_task.cpp [hunk 6]
```diff
@@ -205,12 +212,12 @@ void PipelineTask::set_state(PipelineTaskState state) {
 
 std::string PipelineTask::debug_string() const {
     std::stringstream ss;
-    ss << "PipelineTask(" << _index << ")" << get_state_name(_cur_state) << "\nsink: ";
-    ss << _sink->debug_string();
-    ss << "\n operators(from source to root)";
-    for (auto operatr : _operators) {
-        ss << "\n" << operatr->debug_string();
+    ss << "PipelineTask[id = " << _index << ", state = " << get_state_name(_cur_state)
```
Comment (HappenLee) [inline be/src/pipeline/pipeline_task.cpp:215]: better use `fmt::format` to replace `stringstream`

File: be/src/pipeline/pipeline_task.cpp [hunk 7]
```diff
@@ -86,14 +76,21 @@ bool PipelineTask::has_dependency() {
...
 
 Status PipelineTask::open() {
-    if (_sink) {
-        RETURN_IF_ERROR(_sink->open(_state));
-    }
+    auto st = Status::OK();
     for (auto& o : _operators) {
-        RETURN_IF_ERROR(o->open(_state));
+        if (st.is_blocked_by_rf()) {
+            o->open(_state);
```
Comment (HappenLee) [inline be/src/pipeline/pipeline_task.cpp:82]: should RETURN_IF_ERROR, if some opeartor open failed, should as failed as quickly

File: be/src/vec/exec/scan/vscan_node.cpp (hunk 5 above)
Comment (Gabriel39) [reply be/src/vec/exec/scan/vscan_node.cpp:103]: This argument is true by default. I think it will be false for some cases (maybe external table? )

**Review by HappenLee — APPROVED** [review]
LGTM

**Pull request merged**